db/db_test.cc

```diff
@@ -1352,10 +1356,81 @@ TEST_F(DBTest, ApproximateSizesMemTable) {
   start = Key(1050);
   end = Key(1080);
   r = Range(start, end);
-  db_->GetApproximateSizes(&r, 1, &size_with_mt, include_both);
+  db_->GetApproximateSizes(sizeApproxOpt, defaultColumnFamily, &r, 1,
+                           &size_with_mt);
   db_->GetApproximateSizes(&r, 1, &size_without_mt);
   ASSERT_GT(size_with_mt, size_without_mt);
   ASSERT_GT(size_without_mt, 6000);
+
+  // Check that include_memtabtles flag works as expected
+  sizeApproxOpt.include_memtabtles = false;
+  db_->GetApproximateSizes(sizeApproxOpt, defaultColumnFamily, &r, 1, &size);
+  ASSERT_EQ(size, size_without_mt);
+
+  // Check that files_size_error_margin works as expected, when the heuristic
+  // conditions are not met
+  start = Key(1);
+  end = Key(1000 + N - 2);
+  r = Range(start, end);
+  sizeApproxOpt.files_size_error_margin = -1.0;  // disabled
+  db_->GetApproximateSizes(sizeApproxOpt, defaultColumnFamily, &r, 1, &size);
+  uint64_t size2;
+  sizeApproxOpt.files_size_error_margin = 0.5;  // enabled, but not used
+  db_->GetApproximateSizes(sizeApproxOpt, defaultColumnFamily, &r, 1, &size2);
+  ASSERT_EQ(size, size2);
+}
+
+TEST_F(DBTest, ApproximateSizesFilesWithErrorMargin) {
+  Options options = CurrentOptions();
+  options.write_buffer_size = 1024 * 1024;
+  options.compression = kNoCompression;
+  options.create_if_missing = true;
+  options.target_file_size_base = 1024 * 1024;
+  DestroyAndReopen(options);
+  const auto defaultColumnFamily = db_->DefaultColumnFamily();
+
+  const int N = 64000;
+  Random rnd(301);
+  for (int i = 0; i < N; i++) {
+    ASSERT_OK(Put(Key(i), RandomString(&rnd, 1024)));
+  }
+  // Flush everything to files
+  Flush();
+  // Compact the entire key space into the next level
+  db_->CompactRange(CompactRangeOptions(), defaultColumnFamily, nullptr,
+                    nullptr);
+
+  // Write more keys
+  for (int i = N; i < (N + N / 4); i++) {
+    ASSERT_OK(Put(Key(i), RandomString(&rnd, 1024)));
+  }
+  // Flush everything to files again
+  Flush();
+
+  // Wait for compaction to finish
+  ASSERT_OK(dbfull()->TEST_WaitForCompact());
+
+  const std::string start = Key(0);
+  const std::string end = Key(2 * N);
+  const Range r(start, end);
+
+  SizeApproximationOptions sizeApproxOpt;
+  sizeApproxOpt.include_memtabtles = false;
+  sizeApproxOpt.include_files = true;
+  sizeApproxOpt.files_size_error_margin = -1.0;  // disabled
+
+  // Get the precise size without any approximation heuristic
+  uint64_t size;
+  db_->GetApproximateSizes(sizeApproxOpt, defaultColumnFamily, &r, 1, &size);
+  ASSERT_NE(size, 0);
+
+  // Get the size with an approximation heuristic
+  uint64_t size2;
+  const double error_margin = 0.2;
+  sizeApproxOpt.files_size_error_margin = error_margin;
+  db_->GetApproximateSizes(sizeApproxOpt, defaultColumnFamily, &r, 1, &size2);
+  ASSERT_LT(size2, size);
```

Comment:
Why LT, rather than allowing some ratio? The skipped part can be larger or smaller.

Reply:
ok, fixed it.